02-microarray/pathway-analysis_microarray_07_gsea.Rmd

```diff
@@ -0,0 +1,443 @@
+For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+Let's see a preview of the new `dge_df`. 
+
+```{r}
+head(dge_df)
+```
+
+## Perform gene set enrichment analysis (GSEA)
+
+**DRAFT**
+Gene set enrichment testing using `clusterProfiler` is based on... [@clusterProfiler-book].
+
+### Determine our pre-ranked genes list
+
+The `GSEA()` function takes a pre-ranked (sorted) named vector of statistics, where the names in the vector are gene identifiers.
+This is step 1 -- gene-level statistics.
+
+```{r}
+lfc_vector <- dge_df$logFC
+names(lfc_vector) <- dge_df$entrez_id
+lfc_vector <- sort(lfc_vector, decreasing = TRUE)
+```
+
+```{r}
+# Look at first entries of the log2 fold change vector
+head(lfc_vector)
+```
+
+### Run GSEA using the `GSEA()` function
+
+**DRAFT**
+The enrichment score (ES) for a pathway, a pathway-level statistic, is calculated using our gene-level statistics.
+Genes are ranked from most highly positive to most highly negative and weighting them according to their gene-level statistic.
+A running score is calculated by starting with the most highly ranked genes and increasing the score when a gene is in the pathway and decreasing the score when a gene is not in the pathway.
+The ES is the maximum deviation from zero.
+Significance is assessed by generating a null distribution by sampling random gene sets of the same size and an FDR (false discovery rate) value is calculated to account for multiple hypothesis testing.
+([Subramanian 2005](https://doi.org/10.1073/pnas.0506580102); [Korotkevich 2019](https://doi.org/10.1101/060012)).
+
+**REVIEW**
+We can use the `GSEA()` function to perform GSEA with any generic set of gene sets, but there are several functions for using specific, commonly used gene sets (e.g., `gseKEGG()`).
+
+```{r}
+gsea_results <- GSEA(
+  geneList = lfc_vector, # ordered ranked gene list
+  nPerm = 1000, # number of permutations
+  minGSSize = 25, # minimum gene set size
+  maxGSSize = 500, # maximum gene set set
+  pvalueCutoff = 0.05,
+  pAdjustMethod = "BH", # Benjamini-Hochberg correction
+  TERM2GENE = dplyr::select(
+    dr_hallmark_df,
+    gs_name,
+    entrez_gene
+  )
+)
+```
+
+Let's take a look at the results.
+
+```{r}
+head(gsea_results@result %>%
+  dplyr::arrange(dplyr::desc(NES)))
+```
+
+Normalized enrichment scores (NES) are enrichment scores that are scaled to make gene sets that contain different number of genes comparable.
+
+The information we're most likely interested in is in the `results` slot.
+Let's convert this into a data frame that we can write to file.
+
+```{r}
+gsea_result_df <- data.frame(gsea_results@result)
+```
+
+Let's print out a sneak peek of it here and take a look at how many sets do we have that fit our cutoff of `0.1` FDR? 
```

Comment:
This filter step and the one for ORA is meant to show users how to check if they have significant results. If they don't, the plots just show up blank which may be confusing to users. Seeing blank plots probably wouldn't be clear to users.  

If this isn't clear, we should add some wording to make it clear what this step is for. We could explain that blank plots is what happens too?

Reply:
I see. I think my comment about the filter value matching `pvalueCutoff` still stands. The following code more closely aligns with what you are trying to show in my opinion (note that I am familiar with `any()` and `$` notation and our users may be less so; untested) 

```R
any(gsea_result_df$p.adjust < <THRESHOLD>)
```

And I agree that we could say, when we get to the visualization step, if you see blank plots it's because nothing met your `pvalueCutoff` if that is the situation that creates blank plots.

Reply:
I also think it's also helpful to see what the output looks like which the logic above wouldn't show you.

Reply:
Though I do like making it a definitive TRUE/FALSE test.

Reply:
You see some version of the output in 
```R
head(gsea_results@result %>%
  dplyr::arrange(dplyr::desc(NES)))
```

which is a lot to take in - so I think one could simplify that a bit and then use `any()` here.

Reply:
Ah cool. Sounds good then.